### ISSUE TITLE
Use callpattern heuristics to automatically use apply()

### ISSUE THREAD

**Issue by jburse** [jburse/dogelog-moon#34]
We already explored a heuristic to determine whether something
is a function with return value or a function without return value.
Namely we branch into a fallback register when we dont find a 

functor in the call part of the dual trampolin of the interpreter:
```
   list = kb[functor];
   if (list !== undefined) {
       list = list[arity];
   } else {
       list = register(functor);
   }
```
The register does a lookup in document and then in
window, and returns a pair [obj,res], which is used to
form a closure that involves dispatch(obj,res):
```
    if (args.length>0 && is_variable(args[args.length-1])) {
        args.pop();
        let val=res.apply(obj, args);
        if (!unify(goal.args[goal.args.length-1], val))
            return false;
    } else {
        res.apply(obj, args);
    }
```
So in the above the args are the already dereferenced arguments 
of the actual Prolog call. And it is simply checked whether the last 
argument is a variable. This already allows automatically doing stuff like:
```
:- alert('Hello World!').

:- prompt('Whats your name?', X), write(X), nl.
```
But it lacks object oriented navigation of documents etc..
so idea is here that we introduce and internal is_reference()
test, which will also be a Prolog predicate reference/1

as already found in Jekejeke Prolog. And if the first argument
is indeed a reference datatype, populate the first argument
of apply() with it. It might also influence the lookup, but

basically we can make object oriented single displatch.

**Comment by jburse** [comment]
When we conceived the callpattern heuristic, we didn't know
that JavaScript functions have a length attribute. So back
to basics our initial idea:

https://github.com/jburse/dogelog-moon/issues/14
